File: leetcode-problems/2236-maximum-twin-sum-of-a-linked-list/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        l=0
        curr=head
        while curr:
            l+=1
            curr=curr.next
        v=l//2
        tail=head
        for _ in range(v):
            tail=tail.next
        prev =None
        curr=tail
        while curr:
            temp=curr.next
            curr.next=prev
            prev=curr
            curr=temp
        tail=prev
        max=0
        while tail and head:
            sum=tail.val+head.val
            if sum>max: max=sum
            tail=tail.next
            head=head.next
        return max

File: leetcode-problems/2236-maximum-twin-sum-of-a-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int l=1;
        ListNode* tail=new ListNode();
        tail=head;
        while(tail->next!=nullptr){
            l++;
            tail=tail->next;
        }
        //cout<<l<<'\n';
        ListNode* prev=new ListNode();
        tail=head;
        for (int i=0;i<(l/2);i++){
            prev= tail;
            tail=tail->next;
        }
        //cout<<tail->val<<prev->val<<'\n';
        prev->next=nullptr;
        prev=nullptr;
        //cout<<head->next->val<<'\n';
        ListNode* temp=new ListNode();
        temp=tail->next;
        tail->next=nullptr;
        while (temp!=nullptr){
            prev=tail;
            tail=temp;
            temp=temp->next;
            tail->next=prev;
        }
        int max=0;
        //cout<<head->val<<tail->val<<'\n';
        //cout<<l;
        while(head!=nullptr){
            if ((head->val+tail->val)>max){
                max=head->val+tail->val;
            }
            head=head->next;
            tail=tail->next;
        }
        return max;
    }
};
